TraceSourceLogger now takes exception into account(adds it to the log…

rizi opened the pull request:
TraceSourceLogger now takes the exception into account even if the formatter is not null, therefore the exception will be added to the log message.

This pull request fixes #42341

Please let me know if there is a better way to link the PR to the issue.

Br

rizi committed:
TraceSourceLogger now takes exception into account(adds it to the log message) even if formatter is not null

KalleOlaviNiemitalo reviewed (changes requested):
It looks like this appends the exception to the message twice if `formatter == null`. Once would be enough.

The `string.Format` and `CultureInfo.InvariantCulture` references may be unnecessary; string concatenation could suffice. [EventLogLogger does not use those when it appends an Exception to a StringBuilder.](https://github.com/dotnet/runtime/blob/f4a7e3d4762a179c64022fc0b22d49e3d03d0a02/src/libraries/Microsoft.Extensions.Logging.EventLog/src/EventLogLogger.cs#L115-L118) Exception classes typically do not implement `IFormattable` and thus cannot receive an `IFormatProvider` as a parameter; they just use the ambient cultures. On the other hand, exceptions should be rare, and exception-logging code might then not be performance-sensitive.

Should perhaps add a test in [TraceSourceLoggerProviderTest.cs](https://github.com/dotnet/runtime/blob/f4a7e3d4762a179c64022fc0b22d49e3d03d0a02/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerProviderTest.cs), to verify that Exception.Message occurs in the trace message exactly once. That file has `#if NETFRAMEWORK` though, so I am not sure the test class is even compiled for .NET 5.

rizi committed:
Fixed a logical error, that would have added the exception twice to the log message when the formatter would have been null.

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=42571) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=42571) before we can accept your contribution.<br/><br/>:x: rizi [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=42571)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=42571) it.</sub>

rizi commented:
> It looks like this appends the exception to the message twice if `formatter == null`. Once would be enough.
> 
> The `string.Format` and `CultureInfo.InvariantCulture` references may be unnecessary; string concatenation could suffice. [EventLogLogger does not use those when it appends an Exception to a StringBuilder.](https://github.com/dotnet/runtime/blob/f4a7e3d4762a179c64022fc0b22d49e3d03d0a02/src/libraries/Microsoft.Extensions.Logging.EventLog/src/EventLogLogger.cs#L115-L118) Exception classes typically do not implement `IFormattable` and thus cannot receive an `IFormatProvider` as a parameter; they just use the ambient cultures. On the other hand, exceptions should be rare, and exception-logging code might then not be performance-sensitive.
> 
> Should perhaps add a test in [TraceSourceLoggerProviderTest.cs](https://github.com/dotnet/runtime/blob/f4a7e3d4762a179c64022fc0b22d49e3d03d0a02/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerProviderTest.cs), to verify that Exception.Message occurs in the trace message exactly once. That file has `#if NETFRAMEWORK` though, so I am not sure the test class is even compiled for .NET 5.

@KalleOlaviNiemitalo I fixed the logical issue, that the exception would have been added twice to the log message when the formatter is null.
I also removed the "string formatting".

Regarding tests:
I already asked in the corresponding issue where tests for the LogProvider can be found, I think it would not make sense to add a test for the TraceSourceLogProvider, instead there should already be tests for the TraceSourceProvider, and if there are tests I will of course add one (but I can't find the tests/I'm not sure if there any at all).

KalleOlaviNiemitalo commented:
@rizi, there is [TraceSourceLoggerTest.cs](https://github.com/dotnet/runtime/blob/f4a7e3d4762a179c64022fc0b22d49e3d03d0a02/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerTest.cs) in the same directory. [TraceSourceLoggerProviderTest.BufferedConsoleTraceListener](https://github.com/dotnet/runtime/blob/f4a7e3d4762a179c64022fc0b22d49e3d03d0a02/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerProviderTest.cs#L47) would be useful for the new test, though.

rizi commented:
> @rizi, there is [TraceSourceLoggerTest.cs](https://github.com/dotnet/runtime/blob/f4a7e3d4762a179c64022fc0b22d49e3d03d0a02/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerTest.cs) in the same directory. [TraceSourceLoggerProviderTest.BufferedConsoleTraceListener](https://github.com/dotnet/runtime/blob/f4a7e3d4762a179c64022fc0b22d49e3d03d0a02/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerProviderTest.cs#L47) would be useful for the new test, though.

Thx for clarifying, I will add tests in the TraceSourceLoggerTest.cs file.

rizi commented:
> > @rizi, there is [TraceSourceLoggerTest.cs](https://github.com/dotnet/runtime/blob/f4a7e3d4762a179c64022fc0b22d49e3d03d0a02/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerTest.cs) in the same directory. [TraceSourceLoggerProviderTest.BufferedConsoleTraceListener](https://github.com/dotnet/runtime/blob/f4a7e3d4762a179c64022fc0b22d49e3d03d0a02/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerProviderTest.cs#L47) would be useful for the new test, though.
> 
> Thx for clarifying, I will add tests in the TraceSourceLoggerTest.cs file.
> 
> Can someone please help me to compile the Microsoft.Extensions.Logging.sln
> 
> Here is what I did so far:
> 
> * build.cmd clr+libs -rc Release (finished with 0 warnings and 0 errors)
> * build.cmd -vs Microsoft.Extensions.Logging
> 
> As soon as VS 2019 (16.8 preview 3) opens I can't compile the solution or run tests:
> Error:
> 1>------ Build started: Project: TestUtilities, Configuration: Debug Any CPU ------
> Error occurred while restoring NuGet packages: Invalid restore input. Duplicate frameworks found: 'net5.0, net461, net5.0'. Input files: C:\Dev\Git\runtime\src\libraries\Microsoft.Extensions.Logging\tests\Common\Microsoft.Extensions.Logging.Tests.csproj.
> 1>CSC : error CS0006: Metadata file 'C:\Dev\Git\runtime\artifacts\bin\System.Runtime.CompilerServices.Unsafe\net45-Debug\System.Runtime.CompilerServices.Unsafe.dll' could not be found
> 1>Done building project "TestUtilities.csproj" -- FAILED.
> 1>TestUtilities -> C:\Dev\Git\runtime\artifacts\bin\TestUtilities\net5.0-Debug\TestUtilities.dll
> 2>------ Build started: Project: Microsoft.Extensions.Logging.Tests, Configuration: Debug Any CPU ------
> 
> It seems that VS(nuget restore) are confused with $(NetCoreAppCurrent)-Windows_NT;$(NetCoreAppCurrent).
> 
> As there no way make Visual Studio work?

TraceSourceLogger is an internal class, therefore I can't create an instance with new TraceSourceLogger(...);
Using 
```
 var factory = TestLoggerBuilder.Create(builder => builder.AddTraceSource(testSwitch));
 var logger = factory.CreateLogger("Test");
```

KalleOlaviNiemitalo commented:
@rizi, Microsoft.Extensions.Caching.Memory has [InternalsVisibleToAttribute](https://github.com/dotnet/runtime/blob/f4a7e3d4762a179c64022fc0b22d49e3d03d0a02/src/libraries/Microsoft.Extensions.Caching.Memory/src/Properties/InternalsVisibleTo.cs#L6) to let tests access the internals. I guess you can do the same in Microsoft.Extensions.Logging.TraceSource if you need to. Perhaps that would require [SkipUseReferenceAssembly](https://github.com/dotnet/runtime/blob/f4a7e3d4762a179c64022fc0b22d49e3d03d0a02/src/libraries/Microsoft.Extensions.Caching.Memory/tests/Microsoft.Extensions.Caching.Memory.Tests.csproj#L9) (described in <https://github.com/dotnet/runtime/pull/35606#discussion_r459119463>) as well, because the reference assembly would not have the InternalsVisibleToAttribute.

rizi commented:
> @rizi, Microsoft.Extensions.Caching.Memory has [InternalsVisibleToAttribute](https://github.com/dotnet/runtime/blob/f4a7e3d4762a179c64022fc0b22d49e3d03d0a02/src/libraries/Microsoft.Extensions.Caching.Memory/src/Properties/InternalsVisibleTo.cs#L6) to let tests access the internals. I guess you can do the same in Microsoft.Extensions.Logging.TraceSource if you need to. Perhaps that would require [SkipUseReferenceAssembly](https://github.com/dotnet/runtime/blob/f4a7e3d4762a179c64022fc0b22d49e3d03d0a02/src/libraries/Microsoft.Extensions.Caching.Memory/tests/Microsoft.Extensions.Caching.Memory.Tests.csproj#L9) (described in [#35606 (comment)](https://github.com/dotnet/runtime/pull/35606#discussion_r459119463)) as well, because the reference assembly would not have the InternalsVisibleToAttribute.

Thx for the info, I wasn't sure if that's ok.
Is a mock framework (moq, fakeiteasy,..) for the dotnet/runtime in use? Or how do you assert calls that are made against e.q an interface for a object/class under test?

Any idea how I can make the solution work in visual studio, it's really cumbersome to work with vs code and the command line if you are used to have a feature rich ide.

Br

KalleOlaviNiemitalo commented:
To be clear, I don't have any authority in this project. But if it is OK to use InternalsVisibleToAttribute for tests in Microsoft.Extensions.Caching.Memory, I think it will be OK to use that elsewhere as well.

KalleOlaviNiemitalo commented:
> Is a mock framework (moq, fakeiteasy,..) for the dotnet/runtime in use?

The Microsoft.Extensions.Logging.Tests project uses Moq:

<https://github.com/dotnet/runtime/blob/f4a7e3d4762a179c64022fc0b22d49e3d03d0a02/src/libraries/Microsoft.Extensions.Logging/tests/Common/Microsoft.Extensions.Logging.Tests.csproj#L24>

<https://github.com/dotnet/runtime/blob/f4a7e3d4762a179c64022fc0b22d49e3d03d0a02/src/libraries/Microsoft.Extensions.Logging/tests/Common/LoggerFactoryExtensionsTest.cs#L19-L22>

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.TraceSource/src/TraceSourceLogger.cs:
```diff
@@ -33,15 +35,18 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
+            if (exception != null)
             {
-                _traceSource.TraceEvent(GetEventType(logLevel), eventId.Id, message);
+                string exceptionDelimiter = !string.IsNullOrEmpty(message) ? Environment.NewLine : string.Empty;
```
>string exceptionDelimiter = !string.IsNullOrEmpty(message) ? Environment.NewLine : string.Empty; [](start = 16, length = 96)

nit:  
string exceptionDelimiter = string.IsNullOrEmpty(message) ? string.Empty : Environment.NewLine ;

tarekgh commented:
I think for testing, we can just use the public APIs which will exercise this internal code and get the expected results. I am not sure why we need to test the internal classes? if it is really necessary, I would say use reflection to do so.

rizi committed:
Added tests for TraceSourceLogger.

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerTest.cs:
```diff
@@ -51,10 +55,37 @@ public static void MultipleLoggers_IsEnabledReturnsCorrectValue(SourceLevels fir
 
             var logger = factory.CreateLogger("Test");
-
+            
```
nit remove whitespace

rizi replied:
I can remove the whitespaces, but to be clear this is not one of my tests, it was already there before and I didn't check the other tests. Br

rizi committed:
Removed unnecessary white-spaces.

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging.TraceSource/src/TraceSourceLogger.cs:
```diff
@@ -33,15 +35,18 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
-                if (exception != null)
-                {
-                    message += Environment.NewLine + exception;
-                }
```
- [ ] Why have Environment.NewLine as exception delimiter and not just a single whitespace? " ".
Using single whitespace would allow to keep having the log in a single line for those expected like this until now.

(If you compare this with Console logging, you would see that JsonConsoleFormatter and SimpleConsoleFormatter give the option to either have new line or a single whitespace, but SystemdConsoleFormatter uses single whitespace as delimiter). I would vote to have single whitespace here to make it that bit more consistent with console logging.

- [ ] Even though it becomes more readable to add "Error: " in:
```
message += exceptionDelimiter + "Error: " + exception;
```

I don't see this elsewhere, how about we perhaps stick with below instead?
```c#
message += exceptionDelimiter + exception;
```

- [ ] Nit: now that you removed exception null check in else block, how about from:
```c#
            else
            {
                if (exception != null)
                {
                    message += Environment.NewLine + exception;
                }
            }
```
to 
```c#
            else if (exception != null)
            {
                 message += Environment.NewLine + exception;
            }
```

- [ ] Nit: just to keep consistent with surrounding code blocks perhaps you could add:
```diff
            if (string.IsNullOrEmpty(message) && exception == null)
+           {
                return;
+           {
```

maryamariyan replied:
In summary, recommending:

```c#
            if (!IsEnabled(logLevel))
            {
                return;
            }
            string message = string.Empty;
            if (formatter != null)
            {
                message = formatter(state, exception);
            }
            else if (state != null)
            {
                message += state;
            }

            if (exception != null)
            {
                string exceptionDelimiter = string.IsNullOrEmpty(message) ? string.Empty : " ";
                message += exceptionDelimiter + exception.ToString();
            }
            if (!string.IsNullOrEmpty(message))
            {
                _traceSource.TraceEvent(GetEventType(logLevel), eventId.Id, message);
            }
```

And for sample:
```c#
        static void Main(string[] args)
        {
            using var loggerFactory = LoggerFactory.Create(builder =>
            {
                var testSwitch = new SourceSwitch("sourceSwitch", "Logging Sample");
                testSwitch.Level = SourceLevels.Warning;
            builder
                .AddSystemdConsole()
                .AddTraceSource(testSwitch,
                    new TextWriterTraceListener(@"C:\CodeHub\logs\trace.log"));// writer: Console.Out));
            });
            var logger = loggerFactory.CreateLogger("sourceSwitch");
            logger.Log(LogLevel.Warning, new NotImplementedException("NIE exception message"), "message for warning");
        }
```

Would get a single line log, rather than multiple line for a log:
```
sourceSwitch Warning: 0 : message for warning System.NotImplementedException: NIE exception message
```
Which is similar to SystemdConsoleFormatter output.

rizi replied:
hi, thx for your positive feedback, I absolutely agree with you, next time I try to compare my code to similar implementations
br

maryamariyan reviewed (changes requested):
Thanks for the PR @rizi. I have some nit comments and recommendations on exception delimiter portion of code.

rizi committed:
Make the TraceLogger output more "natural" (one line message)  and similar to SystemdConsoleFormatter.

maryamariyan reviewed (approved):
Aside from nit comments, LGTM.

Thanks @rizi

maryamariyan committed:
Update TraceSourceLoggerTest.cs

maryamariyan committed:
Update TraceSourceLoggerTest.cs

maryamariyan commented:
Just applied two nit commits on your PR. (remote extra new lines or whitespaces) and merging. 

Thanks @rizi for your PR submission.
